node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,353 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metered variant of oneshot channels to be able to extract delays caused by delayed responses.
+
+use std::{
+	convert::TryFrom,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicU64, AtomicU8, Ordering},
+		Arc,
+	},
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::{
+	channel::oneshot::{self, Canceled, Cancellation},
+	future::{Fuse, FusedFuture},
+	prelude::*,
+};
+use futures_timer::Delay;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+/// Provides the reason for termination, or [`Self::Unfinished`] if none exists.
+#[derive(Debug, Clone, Copy, IntoPrimitive, TryFromPrimitive, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Reason {
+	Unfinished = 0,
+	Completion = 1,
+	Cancellation = 2,
+	HardTimeout = 3,
+}
+
+/// A peek into delay and state of a `oneshot` channel.
+#[derive(Debug, Clone, Default)]
+pub struct Meter {
+	inner: Arc<AtomicMeterInner>,
+}
+
+/// Inner meter representation.
+#[derive(Debug, Default)]
+pub struct AtomicMeterInner {
+	/// Delay between first poll and completion or cancellation.
+	first_poll_till_end_in_millis: AtomicU64,
+	/// Determine the reason of completion.
+	reason: AtomicU8,
+}
+
+impl Meter {
+	/// Obtain the duration of a finished or canceled
+	/// `oneshot` channel.
+	/// Returns `None` if the channel still exists.
+	pub fn duration(&self) -> Option<Duration> {
+		if self.reason() == Reason::Unfinished {
+			None
+		} else {
+			let millis = self.inner.first_poll_till_end_in_millis.load(Ordering::Acquire);
+			Some(Duration::from_millis(millis))
+		}
+	}
+
+	/// Obtain the reason to the channel termination.
+	pub fn reason(&self) -> Reason {
+		let reason = self.inner.reason.load(Ordering::Acquire);
+		Reason::try_from(reason).expect(
+			"Input is originally a `Termination` before cast, hence recovery always works. qed",
+		)
+	}
+}
+
+/// Create a new pair of `OneshotMetered{Sender,Receiver}`.
+pub fn channel<T>(
+	name: &'static str,
+	soft_timeout: Duration,
+	hard_timeout: Duration,
+) -> (OneshotMeteredSender<T>, OneshotMeteredReceiver<T>) {
```

Comment:
I would stay consistent with standard onshot channes, which just call this `Sender` and `Receiver`. Maybe leave the `Metered`, so `MeteredReceiver`, `MeteredSender` - I would definitely drop the `Oneshot`.